cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/file/FileSetDataset.java

```diff
@@ -251,4 +253,36 @@ public String getOutputFormatClassName() {
   private String getFileSystemPath(Location loc) {
     return loc.toURI().getPath();
   }
+
+  @Override
+  public void onSuccess() throws DataSetException {
+    // nothing needed to do on success
+  }
+
+  @Override
+  public void onFailure() throws DataSetException {
+    Location outputLocation = getOutputLocation();
+    // If there is no output path, it is either using DynamicPartitioner or the job would have failed.
+    // Either way, we can't do much here.
+    if (outputLocation == null) {
+      return;
+    }
+
+    try {
+      if (outputLocation.exists()) {
+        // Only delete the configured output directory, if it is empty.
+        // On Failure, org.apache.hadoop.mapreduce.lib.output.FileOutputFormat will remove files that it wrote,
+        // but it leaves around the directory that it created.
+        // We don't want to unconditionally delete the output directory on failure, because it may have files written
+        // by a different job.
+        if (outputLocation.isDirectory() && outputLocation.list().isEmpty()) {
```

Comment:
No need to do that many if. Just check if `outputLocation.isDirectory()`, then call `outputLocation.delete()`. The delete() will fail if the directory is not empty.